File: src/types/tool.ts
// 工具相关类型定义,镜像 Rust 端数据结构
//
// 重要约定:Rust 侧 serde 默认将字段名转为 snake_case(除非显式 rename_all),
// 因此 TypeScript 类型也采用 snake_case,保持前后端契约一致。
// 见 PRD 08-data-model.md §3.1 与 05-rust-core-engine.md 的 ToolMetadata。

/** 工具分类,与 Rust 侧 ToolCategory enum 对齐(serde rename_all = "snake_case") */
export type ToolCategory =
  | 'formatter'
  | 'encoder'
  | 'generator'
  | 'parser'
  | 'converter'
  | 'comparator';

/** 工具元数据,UI 只读视角。镜像 Rust `ToolMetadata`。 */
export interface ToolMetadata {
  id: string;
  name: string;
  category: ToolCategory;
  /** lucide-react 图标名 */
  icon: string;
  description: string;
  /** 输入参数 JSON Schema,用于动态渲染表单 */
  input_schema: unknown;
  /** 输出参数 JSON Schema(可选) */
  output_schema?: unknown | null;
  /** 工具标签,用于搜索 */
  tags: string[];
  version: string;
  /** 执行超时(秒),null 表示使用默认值 */
  timeout_secs: number | null;
  /** 是否支持流式执行 */
  streaming_supported: boolean;
}

/** 工具输入,镜像 Rust `ToolInput`(`skip_serializing_if` 不影响反序列化) */
export interface ToolInput {
  text?: string;
  file_path?: string;
  params?: Record<string, unknown>;
}

/** 工具输出元信息,镜像 Rust `OutputMeta` */
export interface OutputMeta {
  duration_ms: number;
  input_bytes: number;
  output_bytes: number;
}

/** 警告级别,镜像 Rust `AlertLevel`(serde rename_all = "snake_case") */
export type AlertLevel = 'info' | 'warning' | 'error';

export interface Alert {
  level: AlertLevel;
  message: string;
}

/** 工具输出,镜像 Rust `ToolOutput` */
export interface ToolOutput {
  text: string;
  extra?: unknown;
  meta?: OutputMeta;
  alerts?: Alert[];
}

/**
 * 工具错误,前端视角。
 *
 * Rust 侧 `ToolError` 通过 `#[serde(tag = "kind", content = "detail")]` 序列化,
 * 但跨 IPC 边界后已被 `CommandResponse.ErrorInfo` 包装为 `{ kind, detail, message }`。
 * 这里以前端最终看到的扁平结构为准。
 */
export interface ToolError {
  code: string;
  message: string;
  details?: unknown;
}

/**
 * 流式事件类型,镜像 Rust `StreamEvent`(`#[serde(tag = "type", rename_all = "snake_case")]`)。
 *
 * Rust 端 4 个变体:
 * - `Progress { percent, message }`
 * - `Chunk { text }`
 * - `Done { output }`
 * - `Error { error }`
 *
 * 注意:taskId 不在事件本身,而是由 Tauri 事件 payload 外层携带。
 */
export type StreamEvent =
  | { type: 'progress'; percent: number; message: string }
  | { type: 'chunk'; text: string }
  | { type: 'done'; output: ToolOutput }
  | { type: 'error'; error: ToolError };

/** 工具运行时上下文(UI 视角,仅暴露给 UI 需要的字段) */
export interface ToolContext {
  tool_id: string;
  /** 是否已取消 */
  cancelled: boolean;
}

File: src-tauri/src/core/tool.rs
use async_trait::async_trait;
use futures::stream::BoxStream;
use serde::{Deserialize, Serialize};

use crate::core::context::ToolContext;
use crate::core::error::ToolError;
use crate::core::input::ToolInput;
use crate::core::output::ToolOutput;

/// 所有工具必须实现的核心 trait
///
/// 实现要点:
/// 1. `metadata()` 返回静态描述,在编译期通过 inventory 注册
/// 2. `execute()` 是纯函数式:相同输入 + 相同 context 配置 → 相同输出
/// 3. 禁止在 execute 中调用 Tauri API,所有外部能力通过 `ToolContext` 注入
#[async_trait]
pub trait Tool: Send + Sync {
    fn metadata(&self) -> &'static ToolMetadata;

    async fn execute(&self, input: ToolInput, ctx: &ToolContext) -> Result<ToolOutput, ToolError>;
}

/// 工具的静态元数据
#[derive(Debug, Clone, Serialize)]
pub struct ToolMetadata {
    pub id: &'static str,
    pub name: &'static str,
    pub category: ToolCategory,
    pub icon: &'static str,
    pub description: &'static str,
    pub input_schema: &'static serde_json::Value,
    pub output_schema: Option<&'static serde_json::Value>,
    pub tags: &'static [&'static str],
    pub version: &'static str,
    pub timeout_secs: Option<u32>,
    pub streaming_supported: bool,
}

/// 工具分类
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ToolCategory {
    Formatter,
    Encoder,
    Generator,
    Parser,
    Converter,
    Comparator,
}

/// 流式工具 trait(可选)
///
/// 仅当 `ToolMetadata.streaming_supported` == true 时实现。
/// `execute_stream` 返回异步流,每个 Item 是一个流式事件。
#[async_trait]
pub trait StreamingTool: Send + Sync {
    fn execute_stream(
        &self,
        input: ToolInput,
        ctx: &ToolContext,
    ) -> BoxStream<'static, Result<StreamEvent, ToolError>>;
}

/// 流式事件
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum StreamEvent {
    Progress { percent: u8, message: String },
    Chunk { text: String },
    Done { output: ToolOutput },
    Error { error: ToolError },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_category_formatter_serde() {
        let v = serde_json::to_value(ToolCategory::Formatter).unwrap();
        assert_eq!(v, "formatter");
    }

    #[test]
    fn test_category_encoder_serde() {
        let v = serde_json::to_value(ToolCategory::Encoder).unwrap();
        assert_eq!(v, "encoder");
    }

    #[test]
    fn test_category_generator_serde() {
        let v = serde_json::to_value(ToolCategory::Generator).unwrap();
        assert_eq!(v, "generator");
    }

    #[test]
    fn test_category_parser_serde() {
        let v = serde_json::to_value(ToolCategory::Parser).unwrap();
        assert_eq!(v, "parser");
    }

    #[test]
    fn test_category_converter_serde() {
        let v = serde_json::to_value(ToolCategory::Converter).unwrap();
        assert_eq!(v, "converter");
    }

    #[test]
    fn test_category_comparator_serde() {
        let v = serde_json::to_value(ToolCategory::Comparator).unwrap();
        assert_eq!(v, "comparator");
    }

    #[test]
    fn test_metadata_static_construction() {
        static SCHEMA: serde_json::Value = serde_json::Value::Null;
        static META: ToolMetadata = ToolMetadata {
            id: "test_tool",
            name: "Test Tool",
            category: ToolCategory::Formatter,
            icon: "circle",
            description: "test",
            input_schema: &SCHEMA,
            output_schema: None,
            tags: &["test"],
            version: "0.1.0",
            timeout_secs: Some(5),
            streaming_supported: false,
        };
        assert_eq!(META.id, "test_tool");
        assert_eq!(META.timeout_secs, Some(5));
        assert!(!META.streaming_supported);
    }

    #[test]
    fn test_stream_event_progress_serde() {
        let ev = StreamEvent::Progress {
            percent: 50,
            message: "half done".into(),
        };
        let v = serde_json::to_value(&ev).unwrap();
        assert_eq!(v["type"], "progress");
        assert_eq!(v["percent"], 50);
        assert_eq!(v["message"], "half done");
    }

    #[test]
    fn test_stream_event_chunk_serde() {
        let ev = StreamEvent::Chunk {
            text: "partial".into(),
        };
        let v = serde_json::to_value(&ev).unwrap();
        assert_eq!(v["type"], "chunk");
        assert_eq!(v["text"], "partial");
    }

    #[test]
    fn test_stream_event_done_serde() {
        let ev = StreamEvent::Done {
            output: ToolOutput {
                text: "final".into(),
                ..Default::default()
            },
        };
        let v = serde_json::to_value(&ev).unwrap();
        assert_eq!(v["type"], "done");
        assert_eq!(v["output"]["text"], "final");
    }

    #[test]
    fn test_stream_event_error_serde() {
        let ev = StreamEvent::Error {
            error: ToolError::Internal("oops".into()),
        };
        let v = serde_json::to_value(&ev).unwrap();
        assert_eq!(v["type"], "error");
        assert_eq!(v["error"]["kind"], "internal");
    }
}
